doc/compiler/optimizer/BenefitInliner.md

```diff
@@ -0,0 +1,94 @@
+> 2. **root call ratio** - It estimates how many times this node is visited per the execution of the root node. 
+> 3. **Inlining method summary** - It captures potential optimization opportunities for inlining this method. (Refer to **InliningMethodSummary.cpp** for details)
+> 4. **static benefit** - The indirect benefit (optimizations unlocked) of inlining this method.
+> 5. **budget** - The remaining budget for the descendants of this node.
+> 6. **Bytecode Index** - The call site index of this method.
+
+The **benefit** of inlining a candidate method is calculated as: `Benefit = root call ratio * static benefit`
+
+The **cost** of inlining a candidate method is equal to the size of this method.
+
+The **cost** and **benefit** will be used for solving the inliner packing problem for getting an optimal set of methods to be inlined.
+
+* **IDT** is an abstract program representation that allows us to consistently define the search space for inlining candidates. The parent-child relationship in the **IDT** corresponds to the caller-caller relationship.
+
+* **IDTBuilder** builds an **IDT** starting from the method (root method) being JIT compiled. The **IDT** is built in **DFS** order. `IDTBuilder::buildIDT()` returns a complete **IDT** that can be consumed by `BenefitInliner::inlinerPacking()`
+
+**Process of building the IDT:**
+
+```
+buildIDT() --> buildIDTHelper() --> generateControlFlowGraph() --> performAbstractInterpretation() --> addChild() 
+             |                                                                                              |  
+             |______________________________________________________________________________________________| 
+``` 
+
+Note: `generateControlFlowGraph()` and `performAbstractInterpretation()` need language-specific implementation.
+
+The responsibilities of the **abstract interpretation**:
+> 1. Identifying call sites. Call sites will be added to the **IDT** as **IDTNodes**
+
+> 2. Generating inlining method summaries. (Refer to **InliningMethodSummary.cpp** for details )
+
+> 3. Simulating the program state. Currently, the useful part of the program state is the arguments ( represented by **AbsArguments**) passed to each call site. Combining those arguments with the inlining method summary, we can get the static benefit of inlining this method.
+
+### 4. AbsVisitor.hpp & IDTBuilderVisitor
+* **AbsVisitor** is an interface for the derived classes to define customized callback methods during abstract interpretation. Currently, there is only one method defined in **AbsVisitor** called `AbsVistor::visitCallSite`. **IDTBuilderVisitor** inherits from **AbsVisitor** and defines it own implementation of `visitCallSite`. Whenever we encounter a call site during abstract interpretation, `AbsVistor::visitCallSite` will be invoked. If **AbsVisitor** is an **IDTBuilderVisitor**, `IDTBuilder::addChild` will be called in order to add an **IDTNode** to the **IDT**.
```

Comment:
I'm not sure I understand this description of `AbsVisitor`. What exactly is it for? Maybe a diagram could help?

Reply:
You will see it in the second phase of contribution. One use case of AbsVisitor is during abstract interpretation, if we encounter a callsite, this callsite will be made into an IDT Node and adding to the IDT. To achieve this, IDTBuilderVisitor inherits from AbsVisitor and implements its own _`visitCallsite`_ method